### PR TITLE
feat: [curve/toos-v2] add list chunkserver

### PR DESCRIPTION
**Pull request by Sindweller**
### What problem does this PR solve?

Issue Number: #2036 

Problem Summary:  support list chunkserver command in [curve tool](https://github.com/opencurve/curve/tree/master/tools-v2)

### What is changed and how it works?

What's Changed: add /tools-v2/pkg/cli/command/curvebs/list/chunkserver.go 

How it Works: curve bs list chunkserver

```
root@d6c388ef51ff:/# ./curve bs list chunkserver 
+----+------+-----------+------+-----------+------------+------------+-----------------------------------------------+--------------+-------------+------------------+-----------+
| ID | TYPE |    IP     | PORT | RWSTATUS  | DISKSTATE  | COPYSETNUM |                  MOUNTPOINT                   | DISKCAPACITY |  DISKUSED   | UNHEALTHYCOPYSET |  EXTADDR  |
+----+------+-----------+------+-----------+------------+------------+-----------------------------------------------+--------------+-------------+------------------+-----------+
| 1  | nvme | 127.0.0.1 | 8201 | READWRITE | DISKNORMAL | 100        | local:///curvebs/playground/chunkserver1/data | 39 GiB       | 42140479488 | 0 %              | 127.0.0.1 |
+----+      +           +------+           +            +------------+-----------------------------------------------+--------------+-------------+------------------+           +
| 2  |      |           | 8202 |           |            | 100        | local:///curvebs/playground/chunkserver2/data | 39 GiB       | 42140479488 | 0 %              |           |
+----+      +           +------+           +            +------------+-----------------------------------------------+--------------+-------------+------------------+           +
| 3  |      |           | 8200 |           |            | 100        | local:///curvebs/playground/chunkserver0/data | 39 GiB       | 42140479488 | 0 %              |           |
+----+------+-----------+------+-----------+------------+------------+-----------------------------------------------+--------------+-------------+------------------+-----------+
```

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by wuhongsong** [comment]
good job, feel free to  communicate if you have any questions

**Comment by Sindweller** [comment]
@Cyber-SiKu plz have a look
and,
1. where should I place the /list/chunkserver/copyset.go? it may not be good to put it under /chunkserver but I have no idea about how to handle it.
2. how can I get the value of  `FLAGS_checkCSAlive`?

File: tools-v2/internal/utils/row.go [hunk 1]
```diff
@@ -109,6 +109,17 @@ const (
 	ROW_VERSION         = "version"
 	ROW_ZONE            = "zone"
 
+	ROW_CHUNK_SERVER_ID         = "chunkServerID"
+	ROW_DISK_TYPE               = "diskType"
+	ROW_HOST_IP                 = "hostIP"
```
Comment (Cyber-SiKu) [inline tools-v2/internal/utils/row.go:114]:
```suggestion
	ROW_IP                 = "IP"
```
is better?

File: tools-v2/pkg/cli/command/curvebs/list/server/server.go [hunk 2]
```diff
@@ -168,3 +170,19 @@ func (pCmd *ServerCommand) RunCommand(cmd *cobra.Command, args []string) error {
...
+	listServer.Cmd.SetArgs([]string{"--format", config.FORMAT_NOOUT})
+	listServer.Cmd.SilenceErrors = true
+	config.AlignFlagsValue(caller, listServer.Cmd, []string{
+		config.CURVEBS_MDSADDR,
+	})
+	err := listServer.Cmd.Execute()
+	if err != nil {
+		retErr := cmderror.ErrBsListZone()
+		retErr.Format(err.Error())
+		return nil, retErr
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/server/server.go:185]:
```suggestion
		return listServer.Servers, retErr
```

**Comment by Cyber-SiKu** [comment]
> 



> @Cyber-SiKu plz have a look and,
> 
> 1. where should I place the /list/chunkserver/copyset.go? it may not be good to put it under /chunkserver but I have no idea about how to handle it.
> 2. how can I get the value of  `FLAGS_checkCSAlive`?

1. lsit/copyset is good idea.
2. add a flag check csalive

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+type ChunkServerCommand struct {
+	basecmd.FinalCurveCmd
+	Rpc      *ChunkServerRpc
+	Response *chunk.GetChunkInfoResponse
+}
+
+var _ basecmd.FinalCurveCmdFunc = (*ChunkServerCommand)(nil) // check interface
+
+func (lRpc *ChunkServerRpc) NewRpcClient(cc grpc.ClientConnInterface) {
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/chunkserver.go:57]:
```suggestion
func (cRpc *ChunkServerRpc) NewRpcClient(cc grpc.ClientConnInterface) {
```

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	lRpc.CsClient = chunk.NewChunkServiceClient(cc)
+}
+
+func (lRpc *ChunkServerRpc) Stub_Func(ctx context.Context) (interface{}, error) {
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/chunkserver.go:61]:
```suggestion
func (cRpc *ChunkServerRpc) Stub_Func(ctx context.Context) (interface{}, error) {
```

Comment (Cyber-SiKu) [reply tools-v2/pkg/cli/command/curvebs/check/chunkserver.go:61]:
```suggestion
func (cRpc *ChunkServerRpc) Stub_Func(ctx context.Context) (interface{}, error) {
```

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	return lRpc.CsClient.GetChunkInfo(ctx, lRpc.Request)
+}
+
+func NewChunkServerCommand() *cobra.Command {
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/chunkserver.go:65]: add check.go, and move this file to chunkserver/chunkserver.go

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/copyset.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+func NewCopySetsInChenkServerCommand() *cobra.Command {
+	return NewGetCopySetsInCopySetCommand().Cmd
+}
+
+func NewGetCopySetsInCopySetCommand() *CopySetCommand {
+	lsCmd := &CopySetCommand{
+		FinalCurveCmd: basecmd.FinalCurveCmd{
+			Use:     "copysets",
+			Short:   "get copysets in chunk server",
+			Example: copysetsExample,
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/copyset.go:77]: These do not need to be defined if they are not subcommands of other commands.

File: tools-v2/pkg/cli/command/curvebs/check/chunkserver.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	result, err := basecmd.GetRpcResponse(pCmd.Rpc.Info, pCmd.Rpc)
+	if err.TypeCode() != cmderror.CODE_SUCCESS {
+		pCmd.Error = err
+		pCmd.Result = result
+		return err.ToError()
+	}
+
+	pCmd.Response = result.(*chunk.GetChunkInfoResponse)
+	var errors []*cmderror.CmdError
+
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/chunkserver.go:150]: no output?

**Review by Cyber-SiKu — COMMENTED** [review]
It is recommended to split into two pr:
check-chunkserver and list-chunkserver

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+// Init implements basecmd.FinalCurveCmdFunc
+func (pCmd *ChunkServerCommand) Init(cmd *cobra.Command, args []string) error {
+	mdsAddrs, err := config.GetBsMdsAddrSlice(pCmd.Cmd)
+	if err.TypeCode() != cmderror.CODE_SUCCESS {
+		return err.ToError()
+	}
+
+	timeout := config.GetFlagDuration(pCmd.Cmd, config.RPCTIMEOUT)
+	retrytimes := config.GetFlagInt32(pCmd.Cmd, config.RPCRETRYTIMES)
+	//FLAG_CHECK_CSALIVE = config.GetFlagBool(pCmd.Cmd, config.CURVEBS_CHECK_CSALIVE)
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:107]: rm

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+			unhealthyRatio := 0.0
+			if config.GetFlagBool(pCmd.Cmd, config.CURVEBS_CHECK_CSALIVE) {
+				// 发rpc重置online状态
+				//csAddr := fmt.Sprintf("%s:%d", info.GetHostIp(), info.GetPort())
+				//isOnline := CheckChunkServerOnline(csAddr)
+				//if isOnline {
+				//	SetOnlineState(topology.OnlineState_ONLINE)
+				//} else {
+				//	SetOnlineState(topology.OnlineState_OFFLINE)
+				//}
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:202]: no need?

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+				}
+
+				//if config.GetFlagBool(pCmd.Cmd, config.CURVEBS_CHECK_HEALTH){
+				//	CheckCopysetsOnChunkServer(csID)
+				//	statistics := GetCopysetStatistics()
+				//	unhealthyRatio = statistics.unhealthyRatio
+				//	if FLAG_UNHEALTHY && unhealthyRatio == 0 {
+				//		continue
+				//	}
+				//}
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:225]: no need?

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+			row := make(map[string]string)
+			row[cobrautil.ROW_CHUNK_SERVER_ID] = fmt.Sprintf("%d", info.GetChunkServerID())
+			row[cobrautil.ROW_DISK_TYPE] = info.GetDiskType()
+			row[cobrautil.ROW_IP] = info.GetHostIp()
+			row[cobrautil.ROW_PORT] = fmt.Sprintf("%d", info.GetPort())
+			row[cobrautil.ROW_RW_STATUS] = fmt.Sprintf("%s", info.GetStatus())
+			row[cobrautil.ROW_DISK_STATE] = fmt.Sprintf("%s", info.GetDiskStatus())
+			row[cobrautil.ROW_COPYSET_NUM] = fmt.Sprintf("%d", len(copysets))
+			row[cobrautil.ROW_MOUNTPOINT] = info.GetMountPoint()
+			row[cobrautil.ROW_DISK_CAPACITY] = fmt.Sprintf("%d", info.GetDiskCapacity())
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:248]:
```suggestion
			row[cobrautil.ROW_DISK_CAPACITY] = humanize.IBytes(info.GetDiskCapacity())
```

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+			row[cobrautil.ROW_DISK_USED] = fmt.Sprintf("%d", info.GetDiskUsed())
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:249]:
```suggestion
			row[cobrautil.ROW_DISK_USED] =  info.GetDiskCapacity(info.GetDiskUsed())
```

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+			row[cobrautil.ROW_UNHEALTHY_COPYSET_RATIO] = fmt.Sprintf("%.0f%%", unhealthyRatio)
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:250]:
```suggestion
			row[cobrautil.ROW_UNHEALTHY_COPYSET_RATIO] = fmt.Sprintf("%.0f %%", unhealthyRatio)
```

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+			row[cobrautil.ROW_EXTERNAL_IP] = info.GetExternalIp()
+			rows = append(rows, row)
+		}
+	}
+
+	list := cobrautil.ListMap2ListSortByKeys(rows, pCmd.Header, []string{cobrautil.ROW_DISK_TYPE, cobrautil.ROW_RW_STATUS, cobrautil.ROW_MOUNTPOINT})
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:256]: only sort by merge
```suggestion
	list := cobrautil.ListMap2ListSortByKeys(rows, pCmd.Header, []string{cobrautil.ROW_IP, cobrautil.ROW_CHUNK_SERVER_ID, cobrautil.ROW_DISK_TYPE})
```

File: tools-v2/internal/utils/row.go [hunk 15]
```diff
@@ -111,6 +111,16 @@ const (
 	ROW_VERSION         = "version"
 	ROW_ZONE            = "zone"
 
+	ROW_CHUNK_SERVER_ID         = "id"
```
Comment (Cyber-SiKu) [inline tools-v2/internal/utils/row.go:114]: already defined

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+	topologyServiceClient topology.TopologyServiceClient
+}
+
+var _ basecmd.RpcFunc = (*ListChunkServerRpc)(nil) // check interface
+
+type ChunkServerCommand struct {
+	basecmd.FinalCurveCmd
+	Rpc      []*ListChunkServerRpc
+	Response []*topology.ListChunkServerResponse
+}
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:57]: listchunkserver can be used independently as a command, and some [other pr](https://github.com/opencurve/curve/pull/2463) need its result.
just like [copyset.go](tools-v2/pkg/cli/command/curvebs/query/copyset/copyset.go).

**Comment by Sindweller** [comment]
> Hi, I am trying to use ListChunkServer method in my job, could you please add an interface for others? like [this method](https://github.com/opencurve/curve/blob/master/tools-v2/pkg/cli/command/curvebs/status/copyset/get_copyset_status.go#L132) in the end

I'll done it.

File: tools-v2/pkg/config/bs.go [hunk 17]
```diff
@@ -103,6 +103,16 @@ const (
 	CURVEBS_SCAN                    = "scan"
 	VIPER_CURVEBS_SCAN              = "curvebs.scan"
 	CURVEBS_DEFAULT_SCAN            = true
+	CURVEBS_CHUNK_SERVER_ID         = "chunkserverid"
+	VIPER_CURVEBS_CHUNK_SERVER_ID   = "curvebs.chunkserverid"
+	CURVEBS_CHECK_CSALIVE           = "checkCSAlive"
+	VIPER_CURVEBS_CHECK_CSALIVE     = "curvebs.checkCSAlive"
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/config/bs.go:109]:
```suggestion
	CURVEBS_CHECK_CSALIVE         = "checkalive"
	VIPER_CURVEBS_CHECK_CSALIVE   = "curvebs.checkAlive"
```
is better?

File: tools-v2/pkg/config/bs.go [hunk 18]
```diff
@@ -103,6 +103,16 @@ const (
...
+	CURVEBS_CHECK_HEALTH            = "checkHealth"
+	VIPER_CURVEBS_CHECK_HEALTH      = "curvebs.checkHealth"
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/config/bs.go:111]:
```suggestion
	CURVEBS_CHECK_HEALTH          = "checkhealth"
	VIPER_CURVEBS_CHECK_HEALTH    = "curvebs.checkHealth"
```

File: tools-v2/pkg/config/bs.go [hunk 19]
```diff
@@ -447,6 +462,26 @@ func AddBsOpRequiredFlag(cmd *cobra.Command) {
...
+	AddBsBoolOptionFlag(cmd, CURVEBS_CHECK_HEALTH, "check chunkserver health")
+}
+
+func AddBSCSOfflineFlag(cmd *cobra.Command) {
+	AddBsBoolOptionFlag(cmd, CURVEBS_CS_OFFLINE, "offline")
+}
+
+func AddBSCSUnhealthyFlag(cmd *cobra.Command) {
+	AddBsBoolOptionFlag(cmd, CURVEBS_CS_UNHEALTHY, "unhealthy")
+}
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/config/bs.go:483]: Add optional to the function name, and the optional flag needs to set a default value.

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+	lCmd.Cmd.SilenceUsage = true
+	lCmd.Cmd.SetArgs([]string{"--format", config.FORMAT_NOOUT})
+	err := lCmd.Cmd.Execute()
+	if err != nil {
+		retErr := cmderror.ErrBsListChunkServer()
+		retErr.Format(err.Error())
+		return nil, retErr
+	}
+	return lCmd.ChunkServerInfos, cmderror.ErrSuccess()
+}
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:273]: If it is always true, this function should not be needed, right?

File: tools-v2/README.md [hunk 21]
```diff
@@ -922,6 +923,28 @@ Output:
...
+
+```bash
+curve bs list chunkserver
+```
+
+Output:
+
+```bash        
++---------------+----------+-----------+------+-----------+------------+------------+-----------------------------------------------+--------------+------------+-----------------------+------------+
+| CHUNKSERVERID | DISKTYPE |  HOSTIP   | PORT | RSSTATUS  | DISKSTATE  | COPYSETNUM |                  MOUNTPOINT                   | DISKCAPACITY |  DISKUSED  | UNHEALTHYCOPYSETRATIO | EXTERNALIP |
```
Comment (caoxianfei1) [inline tools-v2/README.md:938]: `RSSTATUS` to `RWSTATUS`?

File: tools-v2/README.md [hunk 22]
```diff
@@ -922,6 +923,28 @@ Output:
 +----------+---------+---------+---------+------------+---------+
 ```
 
+##### list chunkserver
+
+list chunkserver information in curvebs
+
+```bash
+curve bs list chunkserver
```
Comment (Cyber-SiKu) [inline tools-v2/README.md:931]: add in 
```markdown
## Comparison of old and new commands

### curve bs
```

File: tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+			row[cobrautil.ROW_ID] = fmt.Sprintf("%d", info.GetChunkServerID())
+			row[cobrautil.ROW_TYPE] = info.GetDiskType()
+			row[cobrautil.ROW_IP] = info.GetHostIp()
+			row[cobrautil.ROW_PORT] = fmt.Sprintf("%d", info.GetPort())
+			row[cobrautil.ROW_RW_STATUS] = fmt.Sprintf("%s", info.GetStatus())
+			row[cobrautil.ROW_DISK_STATE] = fmt.Sprintf("%s", info.GetDiskStatus())
+			row[cobrautil.ROW_COPYSET_NUM] = fmt.Sprintf("%d", len(copysets))
+			row[cobrautil.ROW_MOUNTPOINT] = info.GetMountPoint()
+			row[cobrautil.ROW_DISK_CAPACITY] = humanize.IBytes(info.GetDiskCapacity())
+			row[cobrautil.ROW_DISK_USED] = fmt.Sprintf("%d", info.GetDiskCapacity())
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/list/chunkserver/chunkserver.go:230]:
```suggestion
			row[cobrautil.ROW_DISK_USED] = humanize.IBytes(info.GetDiskCapacity())
```

**Comment by Cyber-SiKu** [comment]
cicheck

**Review by Cyber-SiKu — APPROVED** [review]

**Comment by Cyber-SiKu** [comment]
cicheck

**Review by caoxianfei1 — APPROVED** [review]

**Pull request merged**